common/expirecontainer.cpp

```diff
@@ -38,24 +38,28 @@ auto ExpireContainerBase::find(const Item& key_item) -> iterator {
 }
 
 void ExpireContainerBase::clear() {
-    for (auto x : ({
-             SCOPED_LOCK(_lock);
-             _list.node = nullptr;
-             Set(std::move(_set));
-         })) {
-        delete x;
+    Set tmp;
+    {
+        SCOPED_LOCK(_lock);
+        _list.node = nullptr;
+        tmp = std::move(_set);
     }
+    for (auto x : tmp)
+        delete x;
 }
 
 uint64_t ExpireContainerBase::expire() {
-    ({
+    if (_lifespan > UINT64_MAX / 2)
+        return 0;
+    {
         SCOPED_LOCK(_lock);
         _list.split_by_predicate([&](Item* x) {
             bool ret = x->_timeout.expiration() < photon::now;
             if (ret) _set.erase(x);
             return ret;
         });
-    }).delete_all();
+    }
+    _list.delete_all();
     return 0;
```

Comment:
I think the so called 'Statements and Declarations in Expressions' brings nothing but vagueness, and it's not c++ standard.

For example

```
struct A {
    void f () {    
        cout << this << endl;
    }
};
int main() {
    A a;
    cout << &a << endl;
    ({a;}).f();
    // Address not same
}
```

Reply:
It called "Compound Statement Expressions" or "Statement Expression", and it is a long history feature in almost every C/C++ compiler such as GCC/Clang/MSVC/Intel CC/IBM Compiler...  It is a kind of "De Facto Standard"

Statement expression is widely used in a lot of famous projects like
Linux kernel
https://github.com/torvalds/linux/blob/5bbd9b249880dba032bffa002dd9cd12cd5af09c/drivers/most/core.c#L89

SPDK
https://github.com/spdk/spdk/blob/a44a9620a68f2a8c29db1b7165d67fe7c070458c/lib/env_ocf/ocf_env.h#L65

liburing
https://github.com/axboe/liburing/blob/7b3245583069bd481190c9da18f22e9fc8c3a805/src/lib.h#L32

It is hard to say "brings nothing but vagueness".

Since `({a;})` means a expression, the result have to be a pr-value (temporary object) with type `A`, it will be a new object differ from `a`.

A purely C++ way is using inline lambda evaluation instead of statement expression. The code in example should basically equivalent to `[&]{return a;}().f();`, or with less type deducing, `[&]()->A {return a;}().f();`

Reply:
Statement expression is a important in Photon ALog, and cannot be substitute by lambda.

https://github.com/alibaba/PhotonLibOS/blob/7c597803832bef0a11b20ba9b8e5f2aa41b539a9/common/alog.h#L462